cli/src/commands/sign.rs

```diff
@@ -0,0 +1,144 @@
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use clap_complete::ArgValueCandidates;
+use indexmap::IndexSet;
+use itertools::Itertools;
+use jj_lib::commit::Commit;
+use jj_lib::commit::CommitIteratorExt;
+use jj_lib::signing::SignBehavior;
+
+use crate::cli_util::CommandHelper;
+use crate::cli_util::RevisionArg;
+use crate::command_error::user_error_with_hint;
+use crate::command_error::CommandError;
+use crate::complete;
+use crate::ui::Ui;
+
+/// Cryptographically sign a revision
+///
+/// Note that revisions, which are already signed by you, will be resigned.
+///
+/// While that leads to discomfort for users, which sign with hardware devices,
+/// as of now we cannot reliably check if a commit is already signed by the user
+/// without creating a signature (see https://github.com/jj-vcs/jj/issues/5786).
+#[derive(clap::Args, Clone, Debug)]
+pub struct SignArgs {
+    /// What key to use, depends on the configured signing backend.
+    #[arg()]
+    key: Option<String>,
+    /// What revision(s) to sign
+    #[arg(
+        long, short,
+        value_name = "REVSETS",
+        add = ArgValueCandidates::new(complete::mutable_revisions),
+    )]
+    revisions: Vec<RevisionArg>,
+}
+
+pub fn cmd_sign(ui: &mut Ui, command: &CommandHelper, args: &SignArgs) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+
+    if command.settings().signing_backend()?.is_none() {
+        return Err(user_error_with_hint(
+            "No signing backend configured.",
+            "For configuring a signing backend, see https://jj-vcs.github.io/jj/latest/config/#commit-signing",
+        ));
+    }
+
+    let is_authored_by_me =
+        |commit: &Commit| -> bool { commit.author().email == command.settings().user_email() };
+
+    let to_sign: IndexSet<Commit> = workspace_command
+        .parse_union_revsets(ui, &args.revisions)?
+        .evaluate_to_commits()?
+        .try_collect()?;
+
+    workspace_command.check_rewritable(to_sign.iter().ids())?;
+
+    let mut tx = workspace_command.start_transaction();
+
+    let mut signed_commits = vec![];
+    let mut foreign_commits = vec![];
+    let mut num_reparented = 0;
+
+    tx.repo_mut().transform_descendants(
+        to_sign.iter().ids().cloned().collect_vec(),
+        |rewriter| {
+            let old_commit = rewriter.old_commit().clone();
+            let commit_builder = rewriter.reparent();
+
+            if to_sign.contains(&old_commit) {
+                let new_commit = commit_builder
+                    .set_sign_key(args.key.clone())
+                    .set_sign_behavior(SignBehavior::Own)
```

Comment:
Shouldn't this be `Force`?

Reply:
I used `Force` in the very beginning, but can't recall why I changed it to `Own` anymore.
Will change it back to `Force`.